### ISSUE TITLE
ECAPA-TDNN Retrain

### ISSUE THREAD

**Issue by Lallapallooza** [speechbrain/speechbrain#643]
Hello! I tried to retrain your model and it's got 0.95% EER with score normalization instead of 0.69% EER. The training was done with 4 TITAN X using data-parallel backend on vox1+vox2, batch size was set to 56. All other parameters set to default. The training log can be found below. I tried to test 9-10 epochs, the EER is the same. Any ideas what goes wrong? 

```
epoch: 1, lr: 6.56e-04 - train loss: 6.63 - valid loss: 1.81, valid ErrorRate: 5.57e-02                                                                                                                            
epoch: 1, lr: 7.16e-04 - train loss: 5.94 - valid loss: 1.68, valid ErrorRate: 5.03e-02                                                                                                                            
epoch: 2, lr: 5.67e-04 - train loss: 2.36 - valid loss: 9.22e-01, valid ErrorRate: 2.64e-02                                                                                                                        
epoch: 3, lr: 1.53e-04 - train loss: 1.40 - valid loss: 5.96e-01, valid ErrorRate: 1.85e-02                                                                                                                        
epoch: 4, lr: 8.69e-04 - train loss: 1.42 - valid loss: 8.39e-01, valid ErrorRate: 2.42e-02                                                                                                                        
epoch: 5, lr: 4.15e-04 - train loss: 1.55 - valid loss: 6.37e-01, valid ErrorRate: 1.96e-02                                                                                                                        
epoch: 6, lr: 3.02e-04 - train loss: 1.03 - valid loss: 5.02e-01, valid ErrorRate: 1.64e-02                                                                                                                        
epoch: 7, lr: 9.82e-04 - train loss: 1.26 - valid loss: 7.46e-01, valid ErrorRate: 2.18e-02                                                                                                                        
epoch: 8, lr: 2.65e-04 - train loss: 1.27 - valid loss: 5.05e-01, valid ErrorRate: 1.68e-02                                                                                                                        
epoch: 9, lr: 4.51e-04 - train loss: 8.83e-01 - valid loss: 4.81e-01, valid ErrorRate: 1.58e-02                                                                                                                    
epoch: 10, lr: 8.33e-04 - train loss: 1.24 - valid loss: 6.77e-01, valid ErrorRate: 2.04e-02 
```

**Comment by mravanelli** [comment]
Hi,
you can take a look into the results that we have achieved here: https://drive.google.com/drive/folders/1-ahC1xeyPinAHp2oAohL-02smNWO41Cc?usp=sharing. You can take a look into the pre-trained model here: https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb
As you can see, the training log is the following:

```
epoch: 1, lr: 3.89e-04 - train loss: 5.63 - valid loss: 1.50, valid ErrorRate: 4.46e-02
epoch: 2, lr: 3.57e-04 - train loss: 2.62 - valid loss: 9.75e-01, valid ErrorRate: 2.77e-02
epoch: 3, lr: 3.20e-04 - train loss: 1.66 - valid loss: 7.63e-01, valid ErrorRate: 2.29e-02
epoch: 4, lr: 4.26e-04 - train loss: 1.80 - valid loss: 7.96e-01, valid ErrorRate: 2.33e-02
epoch: 5, lr: 8.27e-04 - train loss: 1.33 - valid loss: 8.03e-01, valid ErrorRate: 2.33e-02
epoch: 6, lr: 8.15e-04 - train loss: 1.50 - valid loss: 9.07e-01, valid ErrorRate: 2.61e-02
epoch: 7, lr: 4.39e-04 - train loss: 1.26 - valid loss: 5.65e-01, valid ErrorRate: 1.79e-02
epoch: 8, lr: 9.65e-04 - train loss: 1.30 - valid loss: 7.83e-01, valid ErrorRate: 2.25e-02
epoch: 9, lr: 2.18e-04 - train loss: 1.29 - valid loss: 4.66e-01, valid ErrorRate: 1.60e-02
epoch: 10, lr: 9.88e-04 - train loss: 1.16 - valid loss: 7.76e-01, valid ErrorRate: 2.25e-02
epoch: 11, lr: 6.56e-04 - train loss: 1.12 - valid loss: 5.93e-01, valid ErrorRate: 1.81e-02
epoch: 12, lr: 9.03e-05 - train loss: 1.36 - valid loss: 4.86e-01, valid ErrorRate: 1.66e-02
```
Your numbers in terms of in terms of Valid Error Rate and valid loss look comparable to mine. 
One think that I noticed is that for this task there is a pretty high variability in the results. One relevant difference that might play a role is that I tried it on a single gpu (NVIDIA V100 32 GB) and I used a batch_size=32. In your case you use 4 gpus with batch_size = 56. If you are using DDP, this means that the total batch size is 56 * 4 = 224 (which is much much higher than mine). I suggest doing more iteration (or training with a global batch size of 32).

**Comment by Lallapallooza** [comment]
I was using the data-parallel so the batch a slightly bigger than yours (56 vs 32). I'll try a few more epochs, that may help I think.

**Comment by mycrazycracy** [comment]
> Hi,
> you can take a look into the results that we have achieved here: https://drive.google.com/drive/folders/1-ahC1xeyPinAHp2oAohL-02smNWO41Cc?usp=sharing. You can take a look into the pre-trained model here: https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb
> As you can see, the training log is the following:
> 
> ```
> epoch: 1, lr: 3.89e-04 - train loss: 5.63 - valid loss: 1.50, valid ErrorRate: 4.46e-02
> epoch: 2, lr: 3.57e-04 - train loss: 2.62 - valid loss: 9.75e-01, valid ErrorRate: 2.77e-02
> epoch: 3, lr: 3.20e-04 - train loss: 1.66 - valid loss: 7.63e-01, valid ErrorRate: 2.29e-02
> epoch: 4, lr: 4.26e-04 - train loss: 1.80 - valid loss: 7.96e-01, valid ErrorRate: 2.33e-02
> epoch: 5, lr: 8.27e-04 - train loss: 1.33 - valid loss: 8.03e-01, valid ErrorRate: 2.33e-02
> epoch: 6, lr: 8.15e-04 - train loss: 1.50 - valid loss: 9.07e-01, valid ErrorRate: 2.61e-02
> epoch: 7, lr: 4.39e-04 - train loss: 1.26 - valid loss: 5.65e-01, valid ErrorRate: 1.79e-02
> epoch: 8, lr: 9.65e-04 - train loss: 1.30 - valid loss: 7.83e-01, valid ErrorRate: 2.25e-02
> epoch: 9, lr: 2.18e-04 - train loss: 1.29 - valid loss: 4.66e-01, valid ErrorRate: 1.60e-02
> epoch: 10, lr: 9.88e-04 - train loss: 1.16 - valid loss: 7.76e-01, valid ErrorRate: 2.25e-02
> epoch: 11, lr: 6.56e-04 - train loss: 1.12 - valid loss: 5.93e-01, valid ErrorRate: 1.81e-02
> epoch: 12, lr: 9.03e-05 - train loss: 1.36 - valid loss: 4.86e-01, valid ErrorRate: 1.66e-02
> ```
> 
> Your numbers in terms of in terms of Valid Error Rate and valid loss look comparable to mine.
> One think that I noticed is that for this task there is a pretty high variability in the results. One relevant difference that might play a role is that I tried it on a single gpu (NVIDIA V100 32 GB) and I used a batch_size=32. In your case you use 4 gpus with batch_size = 56. If you are using DDP, this means that the total batch size is 56 * 4 = 224 (which is much much higher than mine). I suggest doing more iteration (or training with a global batch size of 32).

Hi, I also train ECAPA-TDNN with different training set. I use a single V100 GPU and batch size is 32. I use the default setting in train_ecapa_tdnn.yaml. Here is the result:
--------------------
| Training set | EER |
--------------------
| vox1 + vox2 | 0.82 |
--------------------
| vox2            |  1.11 |
--------------------
The result is slightly worse than the public EER. Any advice here? I wonder if the EER changes when the model is trained multiple times.

**Comment by mravanelli** [comment]
Thanks for sharing your results. For this tasks I also have noticed some variability in the performance when training the model different times. I'm not sure what is the exact cause of that. Maybe this is due to the `CyclicLRScheduler`. We might wanna play a bit with the gamma factor that can be used to decrease the lr during training. In any case, probably 0.1-0.2% of EER variations might be something withing a reasonable range of variability.  @Jenthe what is the variability that you observed with your ECAPA model?

**Comment by jazzseow** [comment]
May I know how long it takes you to train ECAPA-TDNN?
I try to train with 2x RTX2080Ti, data-parallel backend, vox1 + vox2 + CNCeleb (around 8k IDs) and it's been 2 weeks and it's still on epoch 2

**Comment by mycrazycracy** [comment]
> May I know how long it takes you to train ECAPA-TDNN?
> I try to train with 2x RTX2080Ti, data-parallel backend, vox1 + vox2 + CNCeleb (around 8k IDs) and it's been 2 weeks and it's still on epoch 2

I train with one V100, vox1+vox2. It takes around 7 days to finish the training.

**Comment by mravanelli** [comment]
Yes, the same for me. One possible bottleneck could be the data reading pipeline (if not done locally as suggested but on a shared filesystem). 
I never tried this code on multiple-gpus, so maybe there could be a bottleneck there as well.

**Comment by TParcollet** [comment]
2 weeks for 2 epochs sounds unreasonable to me. If it's a r/w bottleneck you can try to increase the number of workers (num_workers=X) on each data_loader_opts. Otherwise, have a look at the usage of your GPUS with nvidia-smi. 2x2080 Ti should be slightly slower than a V100 not that much.

**Comment by mravanelli** [comment]
Right, it could be a data reading bottleneck or a multi-gpu one. I confirm
that in my case (vox1 + vox2) takes 7 days for 10-12 epochs.

**Comment by jazzseow** [comment]
Thanks for the help. I'll update here if solve the issue.

**Comment by Fhrozen** [comment]
There is one thing I observed and is related to the mean_var_norm_emb file. Whether I train the model with xvector or ecapa, the `mean_var_norm_emb.ckpt` file is not generated. In the train_speaker_embeddings.py file there is no declaration neither is used along the training. However, in https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb/tree/main/hyperparams.yaml it is generated and seems to be used (I cannot find the training py file). In the files https://drive.google.com/drive/folders/1-ahC1xeyPinAHp2oAohL-02smNWO41Cc, seems to be generated but never used or declared on ty hyperparams.yaml or the training file. Is this file somehow automatically generated? Because I cannot find where is generated from.

**Comment by TParcollet** [comment]
@mravanelli any news on this issue ?

**Comment by jazzseow** [comment]
@mravanelli @mycrazycracy @TParcollet 
Sorry to disturb once again.
I have found out that the timing issue was due to the SERes2NetBlock, specifically this line (in speechbrain/lobes/models/ECAPA_TDNN.py)
```
xl = []
for layer in self.blocks:
    try:
        x = layer(x, lengths=lengths)
    except TypeError:
        x = layer(x)
    xl.append(x)
```
I timed this, and it takes average 1.16s. May I know your timing on this?

**Comment by beiliu253** [comment]
Hi, @mravanelli 

Since your performance is outstanding on vox1+2 (0.69%), I am wondering if you have tried training on vox2 only.

**Comment by mravanelli** [comment]
Yes,
we got EER=0.82%. It will be reported in one upcoming paper.

**Comment by beiliu253** [comment]
Great! Look forward to the paper!😊
